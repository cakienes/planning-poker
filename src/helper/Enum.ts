export enum UserStoryStatusEnum {
    VOTED = 'VOTED',
    ACTIVE = 'ACTIVE',
    NOT_VOTED = 'NOT_VOTED',
}
