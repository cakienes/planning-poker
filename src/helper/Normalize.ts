const maxLength: any = (max: number) => (value: string) => (value && value.length > max ? value.slice(0, max) : value);

export { maxLength };
