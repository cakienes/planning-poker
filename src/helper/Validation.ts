const required: any = (value: any) => (value ? undefined : 'Please fill');

const onlyNumber: any = (value: any) => {
    return value && value !== '0' && parseInt(value, 10) ? undefined : 'Must be a number';
};

export { required, onlyNumber };
