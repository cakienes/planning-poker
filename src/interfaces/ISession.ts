import IUserStory from './IUserStory';

export default interface ISession {
    id: string;
    sessionName: string;
    numberOfVoters: number;
    userStories: IUserStory[];

    optionalUrl?: string;
    developers: string[];
}
