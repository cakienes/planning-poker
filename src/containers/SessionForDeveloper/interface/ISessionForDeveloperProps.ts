import { RouteComponentProps } from 'react-router';
import { UserTypeEnum } from '../../../helper/Enum';
import ISession from '../../../interfaces/ISession';
import IUserStory from '../../../interfaces/IUserStory';

export default interface ISessionForDeveloperProps extends RouteComponentProps<any> {
    activeUserStory?: IUserStory;
    getSessionsFromLocalStorage: () => void;
    giveStoryPoint: (userType: UserTypeEnum, storyPoint?: number) => void;
    selectedSession?: ISession;
    setDeveloper: (developerId: string) => void;
    setNewDeveloper: (guid: string) => void;
    setSelectedSession: (id: string) => void;
}
