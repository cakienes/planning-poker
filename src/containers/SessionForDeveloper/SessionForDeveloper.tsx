import React from 'react';
import { connect } from 'react-redux';
import { toastr } from 'react-redux-toastr';
import { withRouter } from 'react-router-dom';
import uuid from 'uuid';
import ActiveStory from '../../components/ActiveStory/ActiveStory';
import UserStoryList from '../../components/UserStoryList/UserStoryList';
import { UserTypeEnum } from '../../helper/Enum';
import IGlobalState from '../../interfaces/IGlobalState';
import ISession from '../../interfaces/ISession';
import {
    getSessionsFromLocalStorage,
    giveStoryPoint,
    setDeveloper,
    setNewDeveloper,
    setSelecedSessionByUrl,
    setSelectedSession,
} from '../../redux/session/session.actions';
import { getActiveUserStory, getSelectedSession } from '../../redux/session/session.selectors';
import Layout from '../Layout/Layout';
import ISessionForDeveloperProps from './interface/ISessionForDeveloperProps';

export class SessionForDeveloper extends React.Component<ISessionForDeveloperProps, any> {
    constructor(props: ISessionForDeveloperProps) {
        super(props);
        this.state = {};
    }
    componentWillMount() {
        const {
            match: {
                params: { id, optionalUrl },
            },
        } = this.props;
        if (id) {
            this.props.setSelectedSession(id);
        } else if (optionalUrl) {
            this.props.setSelecedSessionByUrl(optionalUrl);
        }
    }

    componentWillReceiveProps(nextProps: ISessionForDeveloperProps) {
        if (!this.props.selectedSession && nextProps.selectedSession && nextProps.selectedSession.id) {
            this.checkSessionAvailability(nextProps.selectedSession);
        }
    }

    componentDidMount(): any {
        setInterval(this.reRenderComponent, 1000);
    }

    render() {
        const {
            selectedSession,
            activeUserStory,
            match: {
                params: { developerId },
            },
        } = this.props;
        return (
            <Layout>
                <div className="session">
                    <div className="row">
                        <div className="col">
                            <UserStoryList selectedSession={selectedSession} />
                        </div>
                        <div className="col-4">
                            <ActiveStory
                                selectedSession={selectedSession}
                                giveStoryPoint={this.props.giveStoryPoint}
                                activeUserStory={activeUserStory}
                                type={UserTypeEnum.DEVELOPER}
                                developerId={developerId}
                            />
                        </div>
                    </div>
                </div>
            </Layout>
        );
    }

    checkSessionAvailability = (selectedSession: ISession): boolean => {
        const {
            match: {
                params: { developerId },
            },
        } = this.props;
        if (!selectedSession) {
            toastr.error('SESSION NOT FOUND', '');
        } else if (developerId) {
            if (!selectedSession.developers.find((x: string) => x === developerId)) {
                this.props.history.push('/session-error');
            } else {
                this.props.setDeveloper(developerId);
            }
        } else if (!developerId) {
            if (selectedSession.developers.length < selectedSession.numberOfVoters) {
                const guid = uuid();
                this.props.history.replace(`/session/${selectedSession.id}/developer/${guid}`);
                this.props.setNewDeveloper(guid);
                return true;
            } else {
                this.props.history.push('/session-error');
            }
        }
        return false;
    };

    reRenderComponent = (): void => {
        this.props.getSessionsFromLocalStorage();
    };
}

const mapStateToProps = (state: IGlobalState) => ({
    activeUserStory: getActiveUserStory(state),
    selectedSession: getSelectedSession(state),
});

export const mapDispatchToProps = (dispatch: Function) => {
    return {
        setSelectedSession: (id: string) => {
            dispatch(setSelectedSession(id));
        },
        giveStoryPoint: (userType: UserTypeEnum, storyPoint?: number) => {
            dispatch(giveStoryPoint(userType, storyPoint));
        },
        setNewDeveloper: (guid: string) => {
            dispatch(setNewDeveloper(guid));
        },
        setDeveloper: (developerId: string) => {
            dispatch(setDeveloper(developerId));
        },
        getSessionsFromLocalStorage: () => {
            dispatch(getSessionsFromLocalStorage());
        },
        setSelecedSessionByUrl: (optionalUrl: string) => {
            dispatch(setSelecedSessionByUrl(optionalUrl));
        },
    };
};

export default withRouter(
    connect(
        mapStateToProps,
        mapDispatchToProps,
    )(SessionForDeveloper),
);
