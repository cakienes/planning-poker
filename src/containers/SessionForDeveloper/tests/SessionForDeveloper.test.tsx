import React from 'react';
import { toastr } from 'react-redux-toastr';
import { createRenderer, ShallowRenderer } from 'react-test-renderer/shallow';
import uuid from 'uuid';
import { UserTypeEnum } from '../../../helper/Enum';
import ISession from '../../../interfaces/ISession';
import {
    getSessionsFromLocalStorage,
    giveStoryPoint,
    setDeveloper,
    setNewDeveloper,
    setSelecedSessionByUrl,
    setSelectedSession,
} from '../../../redux/session/session.actions';
import ISessionForDeveloperProps from '../interface/ISessionForDeveloperProps';
import { mapDispatchToProps, SessionForDeveloper } from '../SessionForDeveloper';

jest.mock('uuid', () => {
    return jest.fn().mockImplementation(() => {
        return 1;
    });
});

jest.mock('react-redux-toastr', () => {
    return {
        toastr: {
            error: jest.fn(),
        },
    };
});

jest.useFakeTimers();

describe('SessionForDeveloper.tsx', () => {
    let renderer: ShallowRenderer;
    const getSessionsFromLocalStorageMock = jest.fn();
    const giveStoryPointMock = jest.fn();
    const setDeveloperMock = jest.fn();
    const setNewDeveloperMock = jest.fn();
    const setSelecedSessionByUrlMock = jest.fn();
    const setSelectedSessionMock = jest.fn();

    let sessionForDeveloperProps: ISessionForDeveloperProps = {
        history: {} as any,
        location: {} as any,
        match: {
            params: {},
        } as any,
        staticContext: {} as any,
        getSessionsFromLocalStorage: getSessionsFromLocalStorageMock,
        giveStoryPoint: giveStoryPointMock,
        setDeveloper: setDeveloperMock,
        setNewDeveloper: setNewDeveloperMock,
        setSelecedSessionByUrl: setSelecedSessionByUrlMock,
        setSelectedSession: setSelectedSessionMock,
    };

    beforeEach(() => {
        renderer = createRenderer();
        sessionForDeveloperProps = {
            history: {} as any,
            location: {} as any,
            match: {
                params: {},
            } as any,
            staticContext: {} as any,
            getSessionsFromLocalStorage: getSessionsFromLocalStorageMock,
            giveStoryPoint: giveStoryPointMock,
            setDeveloper: setDeveloperMock,
            setNewDeveloper: setNewDeveloperMock,
            setSelecedSessionByUrl: setSelecedSessionByUrlMock,
            setSelectedSession: setSelectedSessionMock,
        };
    });

    it('renders correctly SessionForDeveloper', async () => {
        expect(renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />)).toMatchSnapshot();
    });

    it('componentWillMount of SessionForDeveloper while id is not null', async () => {
        sessionForDeveloperProps = {
            ...sessionForDeveloperProps,
            match: {
                params: {
                    id: 'id',
                },
            } as any,
        };
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        instance.componentWillMount();

        expect(setSelectedSessionMock).toHaveBeenCalledWith('id');
    });

    it('componentWillMount of SessionForDeveloper while optionalUrl is not null', async () => {
        sessionForDeveloperProps = {
            ...sessionForDeveloperProps,
            match: {
                params: {
                    optionalUrl: 'optionalUrl',
                },
            } as any,
        };
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        instance.componentWillMount();

        expect(setSelecedSessionByUrlMock).toHaveBeenCalledWith('optionalUrl');
    });

    it('componentWillReceiveProps of SessionForDeveloper nextProps.selectedSession.id is not null', async () => {
        sessionForDeveloperProps = {
            ...sessionForDeveloperProps,
            match: {
                params: {
                    optionalUrl: 'optionalUrl',
                },
            } as any,
        };
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        instance.checkSessionAvailability = jest.fn();

        let dumNextProps: ISessionForDeveloperProps = Object.assign({}, sessionForDeveloperProps);
        dumNextProps.selectedSession = {
            id: 'id',
            developers: [],
            numberOfVoters: 2,
            sessionName: 'sessionName',
            userStories: [],
        };

        instance.componentWillReceiveProps(dumNextProps);

        expect(instance.checkSessionAvailability).toHaveBeenCalledWith(dumNextProps.selectedSession);
    });

    it('componentDidMount of SessionForDeveloper', async () => {
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        instance.reRenderComponent = jest.fn();

        instance.componentDidMount();

        expect(setInterval).toHaveBeenCalledWith(instance.reRenderComponent, 1000);
    });

    it('checkSessionAvailability  of SessionForDeveloper while selectedSession is null', async () => {
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        instance.checkSessionAvailability(undefined);

        expect(toastr.error).toHaveBeenCalledWith('SESSION NOT FOUND', '');
    });

    it('checkSessionAvailability  of SessionForDeveloper while developerId is not null & not a user', async () => {
        sessionForDeveloperProps = {
            ...sessionForDeveloperProps,
            match: {
                params: {
                    developerId: 'developerId',
                },
            } as any,
            history: {
                push: jest.fn(),
            } as any,
        };
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        const dumSelectedSession: ISession = {
            id: 'id',
            developers: [],
            numberOfVoters: 2,
            sessionName: 'sessionName',
            userStories: [],
        };

        instance.checkSessionAvailability(dumSelectedSession);

        expect(sessionForDeveloperProps.history.push).toHaveBeenCalledWith('/session-error');
    });

    it('checkSessionAvailability  of SessionForDeveloper while developerId is not null & a user', async () => {
        sessionForDeveloperProps = {
            ...sessionForDeveloperProps,
            match: {
                params: {
                    developerId: 'developerId',
                },
            } as any,
            history: {
                push: jest.fn(),
            } as any,
        };
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        const dumSelectedSession: ISession = {
            id: 'id',
            developers: ['developerId'],
            numberOfVoters: 2,
            sessionName: 'sessionName',
            userStories: [],
        };

        instance.checkSessionAvailability(dumSelectedSession);

        expect(setDeveloperMock).toHaveBeenCalledWith('developerId');
    });

    it('checkSessionAvailability  of SessionForDeveloper while length < numberOfVoters ', async () => {
        sessionForDeveloperProps = {
            ...sessionForDeveloperProps,
            history: {
                replace: jest.fn(),
            } as any,
        };
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        const dumSelectedSession: ISession = {
            id: 'id',
            developers: [],
            numberOfVoters: 2,
            sessionName: 'sessionName',
            userStories: [],
        };

        const dumGuid = uuid();
        instance.checkSessionAvailability(dumSelectedSession);

        expect(sessionForDeveloperProps.history.replace).toHaveBeenCalledWith(`/session/id/developer/${dumGuid}`);
        expect(setNewDeveloperMock).toHaveBeenCalledWith(dumGuid);
    });

    it('checkSessionAvailability  of SessionForDeveloper while length < numberOfVoters ', async () => {
        sessionForDeveloperProps = {
            ...sessionForDeveloperProps,
            history: {
                push: jest.fn(),
            } as any,
        };
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        const dumSelectedSession: ISession = {
            id: 'id',
            developers: ['developerId'],
            numberOfVoters: 1,
            sessionName: 'sessionName',
            userStories: [],
        };

        instance.checkSessionAvailability(dumSelectedSession);

        expect(sessionForDeveloperProps.history.push).toHaveBeenCalledWith('/session-error');
    });

    it('reRenderComponent of SessionForDeveloper', async () => {
        renderer.render(<SessionForDeveloper {...sessionForDeveloperProps} />);
        const instance: any = renderer.getMountedInstance();

        instance.reRenderComponent();

        expect(getSessionsFromLocalStorageMock).toHaveBeenCalled();
    });

    describe('mapDispatchToProps', () => {
        it('should dispatch setSelectedSession when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);
            result.setSelectedSession('id');
            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(JSON.stringify(setSelectedSession('id')));
        });

        it('should dispatch getSessionsFromLocalStorage when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);
            result.getSessionsFromLocalStorage();
            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(
                JSON.stringify(getSessionsFromLocalStorage()),
            );
        });

        it('should dispatch giveStoryPoint when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);
            result.giveStoryPoint(UserTypeEnum.DEVELOPER, 5);
            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(
                JSON.stringify(giveStoryPoint(UserTypeEnum.DEVELOPER, 5)),
            );
        });

        it('should dispatch setDeveloper when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);
            result.setDeveloper('developerId');
            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(JSON.stringify(setDeveloper('developerId')));
        });

        it('should dispatch setNewDeveloper when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);
            result.setNewDeveloper('developerId');
            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(
                JSON.stringify(setNewDeveloper('developerId')),
            );
        });

        it('should dispatch setSelecedSessionByUrl when called', () => {
            const dispatchMock = jest.fn();
            const result = mapDispatchToProps(dispatchMock);
            result.setSelecedSessionByUrl('sessionUrl');
            expect(JSON.stringify(dispatchMock.mock.calls[0][0])).toEqual(
                JSON.stringify(setSelecedSessionByUrl('sessionUrl')),
            );
        });
    });
});
