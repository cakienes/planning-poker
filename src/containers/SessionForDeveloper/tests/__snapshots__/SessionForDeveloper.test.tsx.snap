// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionForDeveloper.tsx renders correctly SessionForDeveloper 1`] = `
<Layout>
  <div
    className="session"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <UserStoryList />
      </div>
      <div
        className="col-4"
      >
        <ActiveStory
          giveStoryPoint={[MockFunction]}
          type="DEVELOPER"
        />
      </div>
    </div>
  </div>
</Layout>
`;
